### PR TITLE
MultiHashTracker in Search

### PR DESCRIPTION
**Pull request by leogrim**

### CODE REVIEW

File: kifi-backend/modules/search/app/com/keepit/search/ClickHistoryTracker.scala [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+
+import com.keepit.common.db.Id
+import com.keepit.model._
+import com.keepit.common.store._
+import com.amazonaws.services.s3.AmazonS3
+import com.google.inject.{ImplementedBy, Inject, Singleton}
+import com.keepit.common.store.S3Bucket
+
+case class ClickedURI(id: Id[NormalizedURI])
+case class ClickHistoryBuilder (tableSize: Int, numHashFuncs: Int, minHits: Int) extends MultiHashFilterBuilder[ClickedURI]
```
Comment (leogrim) [inline kifi-backend/modules/search/app/com/keepit/search/ClickHistoryTracker.scala:14]: @ymatsuda 
ClickedURI/BrowsedURI is only meant for safe typing

File: kifi-backend/modules/search/app/com/keepit/search/UserHistoryTracker.scala [hunk 2]
```diff
@@ -0,0 +1,16 @@
...
+import com.keepit.common.akka.SafeFuture
+import play.api.libs.concurrent.Execution.Implicits.defaultContext
+
+@Singleton
+class UserHistoryTracker @Inject() (browsingHistoryTracker: BrowsingHistoryTracker, clickHistoryTracker: ClickHistoryTracker)  {
+
+  def getUserHistory(userId: Id[User]): (Future[MultiHashFilter[BrowsedURI]], Future[MultiHashFilter[ClickedURI]]) =
+    (SafeFuture(browsingHistoryTracker.getMultiHashFilter(userId)),
+      SafeFuture(clickHistoryTracker.getMultiHashFilter(userId)))
+}
```
Comment (leogrim) [inline kifi-backend/modules/search/app/com/keepit/search/UserHistoryTracker.scala]: @ymatsuda 
For now this is unoptimized.
Initially I had separate caches for both click and browsing history stores. I removed them and will add a common caching layer in this class. What do you think?

**Comment by ymatsuda** [comment]
Nice work!

File: kifi-backend/modules/search/app/com/keepit/search/BrowsingHistoryTracker.scala [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+  }
+
+  override def -= (key: Id[User]) = {
+    super.-=(key)
+    cache.remove(BrowsingHistoryUserIdKey(key))
+    this
+  }
+
+  override def get(id: Id[User]): Option[MultiHashFilter[BrowsedURI]] = cache.getOrElseOpt(BrowsingHistoryUserIdKey(id))(super.get(id))
+}
```
Comment (leogrim) [inline kifi-backend/modules/search/app/com/keepit/search/BrowsingHistoryTracker.scala:45]: @ymatsuda 
Before I push this to production, I'm worried about possible corruption if we use in-memory caching and I guess we will. What do you think?

**Pull request merged**